File: src/providers/serpapi.ts
import { SearchOptions, SearchProvider, SearchResult, ProviderConfig } from '../types';
import { buildUrl, get, HttpError } from '../utils/http';
import { debug } from '../utils/debug';

/**
 * SerpAPI response types for Google search engine
 */
interface SerpApiSearchResult {
  position: number;
  title: string;
  link: string;
  displayed_link: string;
  snippet: string;
  snippet_highlighted_words?: string[];
  cached_page_link?: string;
  related_pages_link?: string;
  source?: string;
  date?: string;
}

interface SerpApiResponse {
  search_metadata: {
    id: string;
    status: string;
    json_endpoint: string;
    created_at: string;
    processed_at: string;
    google_url: string;
    raw_html_file: string;
    total_time_taken: number;
  };
  search_parameters: {
    engine: string;
    q: string;
    google_domain: string;
    device: string;
    num: number;
    start?: number;
    hl?: string;
    gl?: string;
    safe?: string;
  };
  search_information: {
    organic_results_state: string;
    total_results: number;
    time_taken_displayed: number;
    query_displayed: string;
  };
  organic_results: SerpApiSearchResult[];
  error?: string;
}

/**
 * SerpAPI configuration options
 */
export interface SerpApiConfig extends ProviderConfig {
  /** Search engine to use (e.g., google, bing, yahoo) */
  engine?: string;
  /** Base URL for SerpAPI */
  baseUrl?: string;
}

/**
 * Default base URL for SerpAPI
 */
const DEFAULT_BASE_URL = 'https://serpapi.com/search.json';

/**
 * Creates a SerpAPI provider instance
 * 
 * @param config Configuration options for SerpAPI
 * @returns A configured SerpAPI provider
 */
export function createSerpApiProvider(config: SerpApiConfig): SearchProvider {
  if (!config.apiKey) {
    throw new Error('SerpAPI requires an API key');
  }
  
  const baseUrl = config.baseUrl || DEFAULT_BASE_URL;
  const engine = config.engine || 'google';
  
  return {
    name: 'serpapi',
    config,
    search: async (options: SearchOptions): Promise<SearchResult[]> => {
      const { query, maxResults = 10, page = 1, language, region, safeSearch, timeout, debug: debugOptions } = options;
      
      // Map SDK parameters to SerpAPI parameters
      const params: Record<string, string | number | boolean | undefined> = {
        engine,
        api_key: config.apiKey,
        q: query,
        num: maxResults,
        start: page > 1 ? (page - 1) * maxResults + 1 : undefined,
      };
      
      // Add optional parameters
      if (language) {
        params.hl = language; // Interface language
      }
      
      if (region) {
        params.gl = region; // Country/region
      }
      
      if (safeSearch) {
        params.safe = safeSearch;
      }
      
      const url = buildUrl(baseUrl, params);
      
      // Log request details if debugging is enabled
      debug.logRequest(debugOptions, 'SerpAPI request', {
        url: config.apiKey ? url.replace(config.apiKey, '***') : url,
        params: {
          ...params,
          api_key: '***',
        },
      });
      
      try {
        const response = await get<SerpApiResponse>(url, { timeout });
        
        // Log response if debugging is enabled
        debug.logResponse(debugOptions, 'SerpAPI raw response', {
          status: response.error ? 'error' : 'success',
          itemCount: response.organic_results?.length || 0,
          totalResults: response.search_information?.total_results || 0,
          metadata: response.search_metadata,
        });
        
        if (response.error) {
          throw new Error(`SerpAPI error: ${response.error}`);
        }
        
        if (!response.organic_results || response.organic_results.length === 0) {
          debug.log(debugOptions, 'SerpAPI returned no results');
          return [];
        }
        
        // Transform SerpAPI response to standard SearchResult format
        return response.organic_results.map((result) => {
          // Extract domain from displayed_link
          const domain = result.displayed_link?.split('/')[0];
          
          return {
            url: result.link,
            title: result.title,
            snippet: result.snippet,
            domain,
            publishedDate: result.date,
            provider: 'serpapi',
            raw: result,
          };
        });
      } catch (error) {
        // Create detailed error message with diagnostic information
        let errorMessage = 'SerpAPI search failed';
        let diagnosticInfo = '';
        
        if (error instanceof HttpError) {
          // Handle specific SerpAPI error codes
          if (error.statusCode === 401 || error.statusCode === 403) {
            diagnosticInfo = 'Invalid API key or unauthorized access. Check your SerpAPI key.';
          } else if (error.statusCode === 429) {
            diagnosticInfo = 'Rate limit exceeded. You have reached your SerpAPI usage limit. Check your subscription plan.';
          } else if (error.statusCode === 400) {
            diagnosticInfo = 'Bad request. Check your search parameters, especially the engine value.';
            
            // Try to extract more detailed error info
            if (error.message.includes('parameter is missing')) {
              diagnosticInfo += ' A required parameter is missing from your request.';
            }
          } else if (error.statusCode >= 500) {
            diagnosticInfo = 'SerpAPI server error. The service might be experiencing issues. Try again later.';
          }
          
          errorMessage = `${errorMessage}: ${error.message}`;
        } else if (error instanceof Error) {
          errorMessage = `${errorMessage}: ${error.message}`;
          
          // Check for common error messages
          if (error.message.includes('API key')) {
            diagnosticInfo = 'Invalid or missing API key. Check your SerpAPI key.';
          } else if (error.message.includes('quota') || error.message.includes('limit')) {
            diagnosticInfo = 'You have reached your SerpAPI usage limit. Check your subscription plan.';
          }
        } else {
          errorMessage = `${errorMessage}: ${String(error)}`;
        }
        
        // Add diagnostic info if available
        if (diagnosticInfo) {
          errorMessage = `${errorMessage}\n\nDiagnostic information: ${diagnosticInfo}\n\nSerpAPI docs: https://serpapi.com/search-api`;
        }
        
        // Log detailed error information if debugging is enabled
        debug.log(debugOptions, 'SerpAPI error', {
          error: error instanceof Error ? error.message : String(error),
          statusCode: error instanceof HttpError ? error.statusCode : undefined,
          diagnosticInfo,
        });
        
        throw new Error(errorMessage);
      }
    },
  };
}

/**
 * Pre-configured SerpAPI provider
 * Note: You must call configure before using this provider
 */
export const serpapi = {
  name: 'serpapi',
  config: { apiKey: '' },
  
  /**
   * Configure the SerpAPI provider with your API credentials
   * 
   * @param config SerpAPI configuration
   * @returns Configured SerpAPI provider
   */
  configure: (config: SerpApiConfig) => createSerpApiProvider(config),
  
  /**
   * Search implementation that ensures provider is properly configured before use
   */
  search: async (_options: SearchOptions): Promise<SearchResult[]> => {
    throw new Error('SerpAPI provider must be configured before use. Call serpapi.configure() first.');
  }
};

File: src/providers/serpapi.rs
//! SerpAPI provider

use crate::{
    error::{SearchError, SearchResult},
    types::{SearchOptions, SearchProvider, SearchResult as SearchResultType},
};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Deserialize, Serialize)]
struct SerpApiSearchResult {
    position: Option<u32>,
    title: String,
    link: String,
    displayed_link: Option<String>,
    snippet: Option<String>,
    snippet_highlighted_words: Option<Vec<String>>,
    date: Option<String>,
}

#[derive(Debug, Deserialize)]
struct SerpApiSearchMetadata {
    id: String,
    status: String,
    created_at: String,
    processed_at: String,
    total_time_taken: Option<f32>,
}

#[derive(Debug, Deserialize)]
struct SerpApiSearchInformation {
    total_results: Option<u64>,
    time_taken_displayed: Option<f32>,
    query_displayed: String,
}

#[derive(Debug, Deserialize)]
struct SerpApiResponse {
    search_metadata: Option<SerpApiSearchMetadata>,
    search_information: Option<SerpApiSearchInformation>,
    organic_results: Option<Vec<SerpApiSearchResult>>,
    error: Option<String>,
}

#[derive(Debug)]
pub struct SerpApiProvider {
    api_key: String,
    engine: String,
    base_url: String,
}

impl SerpApiProvider {
    pub fn new(api_key: &str) -> SearchResult<Self> {
        if api_key.is_empty() {
            return Err(SearchError::ConfigError(
                "SerpAPI key is required".to_string(),
            ));
        }

        Ok(Self {
            api_key: api_key.to_string(),
            engine: "google".to_string(),
            base_url: "https://serpapi.com/search.json".to_string(),
        })
    }

    pub fn with_engine(mut self, engine: &str) -> Self {
        self.engine = engine.to_string();
        self
    }

    pub fn with_base_url(mut self, base_url: &str) -> Self {
        self.base_url = base_url.to_string();
        self
    }
}

#[async_trait::async_trait]
impl SearchProvider for SerpApiProvider {
    fn name(&self) -> &str {
        "serpapi"
    }

    async fn search(&self, options: &SearchOptions) -> SearchResult<Vec<SearchResultType>> {
        let client = reqwest::Client::new();

        // Build query parameters with owned strings
        let mut params = HashMap::new();
        params.insert("engine".to_string(), self.engine.clone());
        params.insert("api_key".to_string(), self.api_key.clone());
        params.insert("q".to_string(), options.query.clone());
        params.insert(
            "num".to_string(),
            options.max_results.unwrap_or(10).to_string(),
        );

        // Add pagination if page > 1
        if let Some(page) = options.page {
            if page > 1 {
                let max_results = options.max_results.unwrap_or(10);
                let start = (page - 1) * max_results + 1;
                params.insert("start".to_string(), start.to_string());
            }
        }

        // Add language if specified
        if let Some(ref language) = options.language {
            params.insert("hl".to_string(), language.clone());
        }

        // Add region if specified
        if let Some(ref region) = options.region {
            params.insert("gl".to_string(), region.clone());
        }

        // Add safe search if specified
        if let Some(ref safe_search) = options.safe_search {
            params.insert("safe".to_string(), safe_search.to_string());
        }

        // Make the request
        let request = client.get(&self.base_url).query(&params);

        let response = request.send().await.map_err(|e| SearchError::HttpError {
            message: format!("Failed to send request: {e}"),
            status_code: None,
            response_body: None,
        })?;

        let status = response.status();
        let response_text = response.text().await.map_err(|e| SearchError::HttpError {
            message: format!("Failed to read response: {e}"),
            status_code: None,
            response_body: None,
        })?;

        if !status.is_success() {
            return Err(SearchError::ProviderError(format!(
                "SerpAPI request failed with status {status}: {response_text}"
            )));
        }

        // Parse JSON response
        let serp_response: SerpApiResponse = serde_json::from_str(&response_text).map_err(|e| {
            SearchError::ParseError(format!("Failed to parse SerpAPI response: {e}"))
        })?;

        // Check for API error
        if let Some(error) = serp_response.error {
            return Err(SearchError::ProviderError(format!(
                "SerpAPI error: {error}"
            )));
        }

        // Extract results
        let organic_results = serp_response.organic_results.unwrap_or_default();

        if organic_results.is_empty() {
            return Ok(vec![]);
        }

        // Transform to standard format
        let results = organic_results
            .into_iter()
            .map(|result| {
                // Extract domain from displayed_link or link
                let domain = result
                    .displayed_link
                    .as_ref()
                    .or(Some(&result.link))
                    .and_then(|link| {
                        if let Ok(url) = url::Url::parse(link) {
                            url.host_str().map(|s| s.to_string())
                        } else {
                            link.split('/').next().map(|s| s.to_string())
                        }
                    });

                let raw_value = serde_json::to_value(&result).unwrap_or_default();
                SearchResultType {
                    url: result.link,
                    title: result.title,
                    snippet: result.snippet,
                    domain,
                    published_date: result.date,
                    provider: Some("serpapi".to_string()),
                    raw: Some(raw_value),
                }
            })
            .collect();

        Ok(results)
    }

    fn config(&self) -> HashMap<String, String> {
        let mut config = HashMap::new();
        config.insert("api_key".to_string(), "***".to_string());
        config.insert("engine".to_string(), self.engine.clone());
        config.insert("base_url".to_string(), self.base_url.clone());
        config
    }
}
